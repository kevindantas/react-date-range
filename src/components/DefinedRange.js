import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from '../styles';
import { defaultInputRanges, defaultStaticRanges } from '../defaultRanges';
import { rangeShape } from './DayCell';
import {
  InputRanges,
  InputRange,
  InputRangeInput,
  StaticRanges,
  StaticRange,
  StaticRangeLabel,
  DefinedRangesWrapper,
} from './styles/DefinedRange.styled';

class DefinedRanges extends Component {
  constructor(props) {
    super(props);
    this.state = {
      rangeOffset: 0,
      focusedInput: -1,
    };
    this.handleRangeChange = this.handleRangeChange.bind(this);
  }

  handleRangeChange(range) {
    const { onChange, ranges, focusedRange } = this.props;
    const selectedRange = ranges[focusedRange[0]];
    if (!onChange || !selectedRange) return;
    onChange({
      [selectedRange.key || `range${focusedRange[0] + 1}`]: { ...selectedRange, ...range },
    });
  }

  getSelectedRange(ranges, staticRange) {
    const focusedRangeIndex = ranges.findIndex(range => {
      if (!range.startDate || !range.endDate || range.disabled) return false;
      return staticRange.isSelected(range);
    });
    const selectedRange = ranges[focusedRangeIndex];
    return { selectedRange, focusedRangeIndex };
  }

  render() {
    const { onPreviewChange, ranges, renderStaticRangeLabel, rangeColors, className } = this.props;

    return (
      <DefinedRangesWrapper className={className}>
        {this.props.headerContent}
        <StaticRanges>
          {this.props.staticRanges.map((staticRange, i) => {
            const { selectedRange, focusedRangeIndex } = this.getSelectedRange(ranges, staticRange);
            let labelContent;

            if (staticRange.hasCustomRendering) {
              labelContent = renderStaticRangeLabel(staticRange);
            } else {
              labelContent = staticRange.label;
            }

            return (
              <StaticRange
                type="button"
                selected={Boolean(selectedRange)}
                style={{
                  color: selectedRange
                    ? selectedRange.color || rangeColors[focusedRangeIndex]
                    : null,
                }}
                key={i}
                onClick={() => this.handleRangeChange(staticRange.range(this.props))}
                onFocus={() => onPreviewChange && onPreviewChange(staticRange.range(this.props))}
                onMouseOver={() =>
                  onPreviewChange && onPreviewChange(staticRange.range(this.props))
                }
                onMouseLeave={() => {
                  this.props.onPreviewChange && this.props.onPreviewChange();
                }}>
                <StaticRangeLabel tabIndex={-1}>{labelContent}</StaticRangeLabel>
              </StaticRange>
            );
          })}
        </StaticRanges>
        <InputRanges>
          {this.props.inputRanges.map((rangeOption, i) => (
            <InputRange key={i}>
              <InputRangeInput
                onFocus={() => this.setState({ focusedInput: i, rangeOffset: 0 })}
                onBlur={() => this.setState({ rangeOffset: 0 })}
                onChange={e => {
                  let value = parseInt(e.target.value, 10);
                  value = isNaN(value) ? 0 : Math.max(Math.min(99999, value), 0);
                  this.handleRangeChange(rangeOption.range(value, this.props));
                }}
                min={0}
                max={99999}
                value={
                  rangeOption.getCurrentValue
                    ? rangeOption.getCurrentValue(ranges[this.props.focusedRange[0]] || {})
                    : '-'
                }
              />
              <span className={styles.inputRangeLabel}>{rangeOption.label}</span>
            </InputRange>
          ))}
        </InputRanges>
        {this.props.footerContent}
      </DefinedRangesWrapper>
    );
  }
}

DefinedRanges.propTypes = {
  inputRanges: PropTypes.array,
  staticRanges: PropTypes.array,
  ranges: PropTypes.arrayOf(rangeShape),
  focusedRange: PropTypes.arrayOf(PropTypes.number),
  onPreviewChange: PropTypes.func,
  onChange: PropTypes.func,
  footerContent: PropTypes.any,
  headerContent: PropTypes.any,
  rangeColors: PropTypes.arrayOf(PropTypes.string),
  className: PropTypes.string,
  renderStaticRangeLabel: PropTypes.func,
};

DefinedRanges.defaultProps = {
  inputRanges: defaultInputRanges,
  staticRanges: defaultStaticRanges,
  ranges: [],
  rangeColors: ['#3d91ff', '#3ecf8e', '#fed14c'],
  focusedRange: [0, 0],
};

export default DefinedRanges;
