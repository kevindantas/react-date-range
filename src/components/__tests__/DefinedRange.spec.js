import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import { defaultStaticRanges, defaultInputRanges } from '../../defaultRanges';
import { DefinedRange } from '../..';
import { startOfDay, isSameDay } from 'date-fns';
import { staticRangeSelected } from '../styles/DefinedRange.styled';

describe('The DefinedRange Component', () => {
  const customStaticRanges = [
    {
      label: 'Custom Today',
      range: () => ({
        startDate: startOfDay(new Date()),
        endDate: startOfDay(new Date()),
      }),
      isSelected: () => true,
    },
  ];
  const ranges = [
    {
      startDate: startOfDay(new Date()),
      endDate: startOfDay(new Date()),
      key: 'selection',
    },
  ];

  describe('static ranges', () => {
    it('should list the default static ranges', () => {
      const { getByText } = render(<DefinedRange />);
      defaultStaticRanges.forEach(staticRange => {
        expect(getByText(staticRange.label)).toBeDefined();
      });
    });

    it('should list custom static ranges', () => {
      const { getByText } = render(<DefinedRange staticRanges={customStaticRanges} />);
      customStaticRanges.forEach(staticRange => {
        expect(getByText(staticRange.label)).toBeDefined();
      });
    });

    it('should select a static range', () => {
      let newRange;
      const mockChange = jest.fn().mockImplementation(({ selection }) => {
        newRange = [selection];
      });
      const { getByText, rerender } = render(
        <DefinedRange onChange={mockChange} ranges={ranges} />
      );

      const selectedRange = defaultStaticRanges[1];
      const firstStaticRange = getByText(selectedRange.label);
      fireEvent.click(firstStaticRange);
      expect(mockChange).toHaveBeenCalledWith({
        selection: {
          key: ranges[0].key,
          ...selectedRange.range(),
        },
      });

      rerender(<DefinedRange onChange={mockChange} ranges={newRange} />);
      expect(getByText(selectedRange.label)).toHaveStyleRule(staticRangeSelected);
    });

    it('should call "renderStaticRangeLabel" callback correct amount of times according to the "hasCustomRendering" option', () => {
      const renderStaticRangeLabel = jest.fn();

      render(
        <DefinedRange
          staticRanges={[
            {
              label: 'Dynamic Label',
              range: {},
              isSelected(range) {
                const definedRange = this.range();
                return (
                  isSameDay(range.startDate, definedRange.startDate) &&
                  isSameDay(range.endDate, definedRange.endDate)
                );
              },
              hasCustomRendering: true,
            },
            {
              label: 'Static Label',
              range: {},
              isSelected(range) {
                const definedRange = this.range();
                return (
                  isSameDay(range.startDate, definedRange.startDate) &&
                  isSameDay(range.endDate, definedRange.endDate)
                );
              },
            },
            {
              label: 'Hede',
              range: {},
              isSelected(range) {
                const definedRange = this.range();
                return (
                  isSameDay(range.startDate, definedRange.startDate) &&
                  isSameDay(range.endDate, definedRange.endDate)
                );
              },
              hasCustomRendering: true,
            },
          ]}
          renderStaticRangeLabel={renderStaticRangeLabel}
        />
      );

      expect(renderStaticRangeLabel).toHaveBeenCalledTimes(2);
    });
  });

  describe('input ranges', () => {
    it('should update the selection on input change', () => {
      const mockChange = jest.fn();
      const { getByDisplayValue } = render(<DefinedRange onChange={mockChange} ranges={ranges} />);

      const input = getByDisplayValue((content, elem) => {
        return elem.parentNode.textContent === defaultInputRanges[0].label;
      });

      const days = 2;
      fireEvent.change(input, {
        target: {
          value: days,
        },
      });

      const expectedRange = {
        selection: {
          key: ranges[0].key,
          ...defaultInputRanges[0].range(days),
        },
      };
      expect(mockChange).toHaveBeenCalledWith(expectedRange);
    });
  });

  describe('snapshots', () => {
    it('should render dynamic static label contents correctly', () => {
      const renderItalicLabelContent = () => (
        <i className={'italic-label-content'}>{'Italic Content'}</i>
      );
      const renderBoldLabelContent = () => <b className={'bold-label-content'}>{'Bold Content'}</b>;
      const renderSomethingElse = () => <img className={'random-image'} />;

      const renderStaticRangeLabel = function(staticRange) {
        let result;

        if (staticRange.id === 'italic') {
          result = renderItalicLabelContent();
        } else if (staticRange.id === 'bold') {
          result = renderBoldLabelContent();
        } else {
          result = renderSomethingElse();
        }

        return result;
      };

      const { container } = render(
        <DefinedRange
          staticRanges={[
            {
              id: 'italic',
              range: {},
              isSelected(range) {
                const definedRange = this.range();
                return (
                  isSameDay(range.startDate, definedRange.startDate) &&
                  isSameDay(range.endDate, definedRange.endDate)
                );
              },
              hasCustomRendering: true,
            },
            {
              label: 'Static Label',
              range: {},
              isSelected(range) {
                const definedRange = this.range();
                return (
                  isSameDay(range.startDate, definedRange.startDate) &&
                  isSameDay(range.endDate, definedRange.endDate)
                );
              },
            },
            {
              id: 'whatever',
              range: {},
              isSelected(range) {
                const definedRange = this.range();
                return (
                  isSameDay(range.startDate, definedRange.startDate) &&
                  isSameDay(range.endDate, definedRange.endDate)
                );
              },
              hasCustomRendering: true,
            },
            {
              id: 'bold',
              range: {},
              isSelected(range) {
                const definedRange = this.range();
                return (
                  isSameDay(range.startDate, definedRange.startDate) &&
                  isSameDay(range.endDate, definedRange.endDate)
                );
              },
              hasCustomRendering: true,
            },
          ]}
          renderStaticRangeLabel={renderStaticRangeLabel}
        />
      );

      expect(container).toMatchSnapshot();
    });
  });
});
