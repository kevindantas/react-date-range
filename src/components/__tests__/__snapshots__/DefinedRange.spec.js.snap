// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The DefinedRange Component snapshots should render dynamic static label contents correctly 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  display: block;
  outline: 0;
  line-height: 18px;
  padding: 10px 20px;
  text-align: left;
}

.c2 {
  border: 0;
  cursor: pointer;
  display: block;
  outline: 0;
  border-bottom: 1px solid #eff2f7;
  font-size: inherit;
  padding: 0;
  background: #fff;
  -webkit-transition: background-color 0.2s ease-in;
  transition: background-color 0.2s ease-in;
}

.c2:hover,
.c2:focus {
  background: #eff2f7;
}

.c8 .c3 {
  font-weight: bold;
}

.c0 {
  font-size: 12px;
  width: 226px;
  border-right: solid 1px #eff2f7;
  background: #fff;
}

.c5 {
  padding: 10px 0;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5px 20px;
}

.c7 {
  width: 30px;
  height: 30px;
  line-height: 30px;
  border-radius: 4px;
  text-align: center;
  border: solid 1px rgb(222,231,235);
  margin-right: 10px;
  color: rgb(108,118,122);
}

.c7:focus,
.c7:hover {
  border-color: rgb(180,191,196);
  outline: 0;
  color: #333;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <button
        class="c2"
        type="button"
      >
        <span
          class="c3 c4"
          tabindex="-1"
        >
          <i
            class="italic-label-content"
          >
            Italic Content
          </i>
        </span>
      </button>
      <button
        class="c2"
        type="button"
      >
        <span
          class="c3 c4"
          tabindex="-1"
        >
          Static Label
        </span>
      </button>
      <button
        class="c2"
        type="button"
      >
        <span
          class="c3 c4"
          tabindex="-1"
        >
          <img
            class="random-image"
          />
        </span>
      </button>
      <button
        class="c2"
        type="button"
      >
        <span
          class="c3 c4"
          tabindex="-1"
        >
          <b
            class="bold-label-content"
          >
            Bold Content
          </b>
        </span>
      </button>
    </div>
    <div
      class="c5"
    >
      <label
        class="c6"
      >
        <input
          class="c7"
          max="99999"
          min="0"
          value="-"
        />
        <span>
          days up to today
        </span>
      </label>
      <label
        class="c6"
      >
        <input
          class="c7"
          max="99999"
          min="0"
          value="-"
        />
        <span>
          days starting today
        </span>
      </label>
    </div>
  </div>
</div>
`;
